core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala

```diff
@@ -797,12 +797,26 @@ private[sql] final case class ColumnTableScan(
              |    ? $decoderGlobal.getStringDictionary()
              |    : $mutableDecoderGlobal.getStringDictionary();
           """.stripMargin, s"($dictionaryVar == null)", dictionaryVar)
-        val dictionaryIndex = ExprCode(
-          s"""
-             |final int $dictionaryIndexVar = $updateDecoder == null
-             |    ? $decoder.readDictionaryIndex($buffer, $nonNullPosition)
-             |    : $updateDecoder.readDictionaryIndex();
+        val dictionaryIndex = if (attr.nullable) {
+          ExprCode(
+            s"""
+               |if ($decoder.isNullAt($buffer, $batchOrdinal)) {
+               |  $dictionaryIndexVar = $dictionaryVar.size();
+               |  $numNullsVar++;
+               |} else {
+               |  $dictionaryIndexVar = $updateDecoder == null
+               |    ? $decoder.readDictionaryIndex($buffer, $nonNullPosition)
+               |    : $updateDecoder.readDictionaryIndex();
+               |}
+               """.stripMargin, "false", dictionaryIndexVar)
```

Comment:
Ideally could also initialize the isNullVar to avoid doing twice but can be noted and done later (or I can take care in my set of changes).